src/tls.c

```diff
@@ -910,11 +914,14 @@ static int connTLSWrite(connection *conn_, const void *data, size_t data_len) {
 
     if (conn->c.state != CONN_STATE_CONNECTED) return -1;
     ERR_clear_error();
+    serverAssert(data_len >= conn->last_failed_write_data_len);
```

Comment:
This is triggering in the test because of the periodic write pings that we do. Maybe instead of crashing we return EAGAIN here?

Reply:
Link to the location, https://github.com/valkey-io/valkey/blob/f85c93301ca63f070164b307c6a7371a9ab5ac7b/src/replication.c#L1957. @ranshid and I also discussed a new connection function, "connBestEffortWrite" which could drop the state here.

Reply:
you are right I completely forgot about that one. let me take a look

Reply:
I think it just expose another way to trigger this issue, when some data wasn't written correctly and retry it with something else that may cause invalid protocol data (is sending new line in the middle of the other command safe?).
Returning `EAGAIN` should fix it, but I think not sending this ping when any data is awaiting to be sent is more accurate.
I added assert to expose anything that may cause invalid write to openssl, maybe also change it for now to `debugServerAssert`?

Reply:
I am not sure how we can solve it like this.
For example it is possible that we consumed all the write buffer. in this case it is possible that a followup ping send would be O.K to go (the connection has no last_failed_write_data_len > 0) however it MIGHT be blocked on the ping (return SSL_ERROR_WANT_WRITE). in this case even if we ignore the fact that the ping failed, the next send data we will try to send will miss the first bytes (since the ping is cached inside the openssl). I tested this in a test program and indeed it showed that this would be the case.

Reply:
Yes, you are right, but this seems to be an issue inside ping and probably ping should go via `client` not `connection` directly in https://github.com/valkey-io/valkey/blob/f85c93301ca63f070164b307c6a7371a9ab5ac7b/src/replication.c#L1957
but I don't know is it possible use `server.primary` inside ping method, if I didn't miss anything it should be `client` for that connection

Reply:
This code is on the replica side. I do not think this is the problem. more possible the issue observed in the tests is related to sending pings

Reply:
Actually I am more confused. forgot that for pings (real ping commands we send to the replica) we are feeding the replication backlog so that should not meet this assert. the case of sending OutOfBand keepalive (\n) we do it from the primary only in cases were we should not have anything "pending" on the socket/SSL to the replica since we are either post exchange or we are writing RDB to disk (in which case we should not be streaming to the replica even with RDB second channel). I am probably missing a case were this is not true, but I cannot think of which.

Reply:
@zori-janea I think we need to understand how these ping (if indeed this is the case) are getting mixed in here.
I suggest we do return with the EAGAIN now and investigate the potential issue with the '\n' later.

Reply:
Maybe EAGAIN is not the correct thing to do? I think we should return an error since returning EAGIN is usually done with updateSSLEvent. I think we can return an error like -1 and set the errno to EBUSY or something. In the good path the error should not be taken into any consideration IMO

Reply:
@ranshid is it ok?
```suggestion
    if (data_len >= conn->last_failed_write_data_len) {
        conn->c.last_errno = EBUSY;
        return -1;
    }
```

Whatever we put here should be won't be correct until we find what is trying to send some smaller buffer then previous. We can also just remove this assert and merge only changes in `connTLSWritev`

Reply:
Maybe you are right. removing the assert would probably still leave the same error as we have today for the small writes. I would be fine with that

Reply:
@zori-janea can you please just fix the comment as @xbasel suggested?

Reply:
@ranshid I added it as suggestion, as I'm not very sure how to phrase it correctly
take a look at https://github.com/valkey-io/valkey/pull/1737#discussion_r1961469388 and https://github.com/valkey-io/valkey/pull/1737#discussion_r1961469388

Reply:
@zori-janea I rethoughty - we cannot just remove the assert since it would cause the connections state to be CONN_STATE_ERROR. lets just return the made-up error.

Reply:
I pushed changes, take a look